echo-telemetry/src/main/java/com/netflix/spinnaker/echo/telemetry/TelemetryEventListener.java

```diff
@@ -0,0 +1,168 @@
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.hash.Hashing;
+import com.google.protobuf.util.JsonFormat;
+import com.netflix.spinnaker.echo.config.TelemetryConfig;
+import com.netflix.spinnaker.echo.events.EchoEventListener;
+import com.netflix.spinnaker.echo.model.Event;
+import com.netflix.spinnaker.kork.proto.stats.Application;
+import com.netflix.spinnaker.kork.proto.stats.Execution;
+import com.netflix.spinnaker.kork.proto.stats.SpinnakerInstance;
+import com.netflix.spinnaker.kork.proto.stats.Stage;
+import com.netflix.spinnaker.kork.proto.stats.Status;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+import retrofit.mime.TypedString;
+
+@Slf4j
+@Component
+@ConditionalOnProperty("telemetry.enabled")
+public class TelemetryEventListener implements EchoEventListener {
+
+  private static final Set<String> LOGGABLE_DETAIL_TYPES =
+      ImmutableSet.of(
+          "orca:orchestration:complete",
+          "orca:orchestration:failed",
+          "orca:pipeline:complete",
+          "orca:pipeline:failed");
+
+  private static final JsonFormat.Printer JSON_PRINTER =
+      JsonFormat.printer().omittingInsignificantWhitespace();
+
+  private final TelemetryService telemetryService;
+
+  private final TelemetryConfig.TelemetryConfigProps telemetryConfigProps;
+
+  @Autowired
+  public TelemetryEventListener(
+      TelemetryService telemetryService,
+      TelemetryConfig.TelemetryConfigProps telemetryConfigProps) {
+    this.telemetryService = telemetryService;
+    this.telemetryConfigProps = telemetryConfigProps;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public void processEvent(Event event) {
+    try {
+      if (event.getDetails() == null || event.getContent() == null) {
+        log.debug("Telemetry not sent: Details or content not found in event");
+        return;
+      }
+
+      String eventType = event.getDetails().getType();
+      if (!LOGGABLE_DETAIL_TYPES.contains(eventType)) {
+        log.debug("Telemetry not sent: type '{}' not whitelisted ", eventType);
+        return;
+      }
+
+      String applicationId = event.getDetails().getApplication();
+      if (applicationId == null || applicationId.isEmpty()) {
+        log.debug("Application ID must be non-null and not empty");
+        return;
+      }
+
+      Map<String, Object> execution = (Map<String, Object>) event.getContent().get("execution");
+      if (execution == null || execution.isEmpty()) {
+        log.debug("Missing execution from Event content.");
+        return;
+      }
+
+      String hashedApplicationId = hash(applicationId);
+      String hashedExecutionId = hash(execution.getOrDefault("id", "").toString());
```

Comment:
For a missing execution ID, we're going to the `id` field to sha256("") which seems less useful than just not setting it at all, doesn't it?

Reply:
Done.